### ISSUE TITLE
Query : How to work with variable width input images?

### ISSUE THREAD

**Issue by NightFury13** [Halfish/lstm-ctc-ocr#2]
@Halfish the images of the numeral-sequences you create are all of the same dimensions (36 x 255). How can one go about using variable width images instead? (36 x var-width)

I see that you [feed in a complete image](https://github.com/Halfish/lstm-ctc-ocr/blob/master/3_phonernn.lua#L119) at a time to the model (not as 255-strips of 36pix-height of the image -- a sequence of pixel-strips). Can you suggest how this can be modified to incorporate variable length inputs?

**Comment by NightFury13** [comment]
Update : I think even you are giving the image in strips just that the model automatically takes care of the splitting into multiple timesteps for you? [step(2) of the model].

```
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> output]
  (1): nn.SplitTable
  (2): nn.Sequencer @ nn.FastLSTM(36 -> 128)          <--- this is the 36pix-height strip?
  (3): nn.Sequencer @ nn.Recursor @ nn.ReLU
  (4): nn.Sequencer @ nn.Recursor @ nn.BatchNormalization
  (5): nn.Sequencer @ nn.Recursor @ nn.Dropout(0.5, busy)
  (6): nn.Sequencer @ nn.Recursor @ nn.Linear(128 -> 12)
  (7): nn.JoinTable
  (8): nn.View(255, 12)
}
```

 However, I'm still unable to figure out my original problem. How do I get this to work for inputs of varying width (not fixed at 255)? 

According to the [documentation](https://github.com/Element-Research/rnn#sequencer), _"The Sequencer forwards an input sequence (a table) into an output sequence (a table of the same length)"_ and that _"The Sequencer requires inputs and outputs to be of shape seqlen x batchsize x featsize"_ So does this mean that the _seqlen_ of each sample in the batch has to be the same?

The current solution I see is adding a pre-processing step of indexing all the same width images into buckets and then while training, pick a random bucket and then train on a batch of images of that bucket. Is there a better/lesser-tedious way of doing this?

**Comment by Halfish** [comment]
@NightFury13 I have not tried variable length width yet, since it is a little complex to deal with such situation, and actually this is my first RNN toy project. I will just share some basic ideas as following.
1. To be able to accept variable inputs, the most simple way is just padding your images with white pixels. I have tried this before and it was easy to implement. However, this may limit your image size with a maximum value H and also waste your computing resources.
2. Another way to accept variable width inputs is just training the model without batch. By feeding lstm with input samples one by one, you don't need to worry about the sequence length. But this may requires torch to allocate and release memory frequently if no memory-share-mechanism was applied. I don't know how to do it or whether torch has already done it, you need to google it. Besides, training without batch and abandon batch normalization layer means that the model will converge slowly and the loss will decrease with fluctuations.
3. Finally, as your fore-mentioned, prepare some buckets with different lengths at first, then put the sample into the corresponding bucket(may also need a little padding),  whenever a bucket reach the batch size, you can use this batch bucket to train the model. 

So maybe this 'bucket-way' is the most common way to do such thing, considering training sequence with variable lengths is inherently one tedious work. And I heard tensorflow had a better bucket training support with seq-to-seq model.

**Comment by NightFury13** [comment]
@Halfish thanks for the advice :smile:  With some more searching around it looks like masking the input is a good idea as the criterion ignores those inputs-values while computing errors/gradients, so I'll be trying that next.

As for the the bucket-index method, I feel I've taken care of all data prep and feeding however can't figure out one detail of the last-layer for the model. `model:add(nn.View(decoder.max_steps, decoder.vocab_size))`. In case of the fixed dimensional input, we can preset the max_steps and hence it works fine. For the case of variable-length sequences, we would want to model to modify this parameter for each batch of inputs.

I tried something like `model:add(nn.View(-1, decoder.vocab_size):setNumInputDims(2))` but it doesn't work as the _batchsize_ and _seq-length_ get multiplied and the output vector is 2-dim instead of being 3-dim as expected. I have different number of elements in each of the buckets so I didn't want to fix the batchsize (in which case I could have done `model:add(nn.View(batchsize, -1, decoder.vocab_size):setNumInputDims(3))` and that works (tested). Can you tell me if there is a way of not fixing both the _batchsize_ as well as _sequence-length_? 

Addn : I see that you have the same line in a function of the deocder.lua

```
function decoder_util:outputs2targets(outputs)
    --outputs = nn.View(self.max_steps, self.vocab_size):forward(outputs:double())
```

Is there some way in which I can remove this View-layer from the model and use it only per-batch somehow as in the decoder?

**Comment by Halfish** [comment]
Maybe you can remove the _View layer_, and do this dimension adjustment dynamically in the `training process`? for example, add your code in [this line](https://github.com/Halfish/lstm-ctc-ocr/blob/master/3_phonernn.lua#L124).

**Comment by NightFury13** [comment]
@Halfish : thanks for that trick! Updated my codes, waiting for results. Will update with my findings.

For any future viewers of this thread...here's what I did.

```
        -- forward pass
        local outputs = model:forward(inputs)

        -- convert outputs for CTC
        outputs = nn.View(width, vocab_size):forward(outputs:double())

        -- calc CTC loss
        local loss = ctcCriterion:forward(outputs, targets, torch.Tensor(sizes))
        local gradOutput = ctcCriterion:backward(outputs, targets)

        -- convert back for model
        gradOutput = nn.View(-1, vocab_size):forward(gradOutput)

        -- backward pass

```
